tests/examples/test_examples.py

```diff
@@ -0,0 +1,148 @@
+        # would be nice to implement support later of course
+        if (
+            "This application failed to start because no Qt platform plugin could be initialized."
+            in result.stdout
+        ):
+            pytest.skip("Qt examples are currently not supported on headless")
+        if "ModuleNotFoundError: No module named 'wx'" in result.stdout:
+            pytest.skip("wx library is not available")
+
+    assert zero_exit, f"failed to run:\n{result.stdout}"
+
+
+@pytest.fixture
+def force_offscreen():
+    """Force the offscreen canvas to be selected by the auto gui module."""
+    os.environ["WGPU_FORCE_OFFSCREEN"] = "true"
+    try:
+        yield
+    finally:
+        del os.environ["WGPU_FORCE_OFFSCREEN"]
+
+
+@pytest.fixture
+def mock_time():
+    """Some examples use time to animate. Fix the return value
+    for repeatable output."""
+    with patch("time.time") as time_mock:
+        time_mock.return_value = 1.23456
+        yield
+
+
+@pytest.mark.parametrize("module", examples_to_test)
+def test_examples_screenshots(module, pytestconfig, force_offscreen, mock_time):
+    """Run every example marked for testing."""
+
+    # render
+    example = importlib.import_module(f"examples.{module}")
+    img = example.canvas.draw()
+
+    # check if _something_ was rendered
+    assert img is not None and img.size > 0
+
+    # we skip the rest of the test if you are not using lavapipe
+    # images come out subtly differently when using different wgpu adapters
+    # so for now we only compare screenshots generated with the same adapter (lavapipe)
+    # a benefit of using pytest.skip is that you are still running
+    # the first part of the test everywhere else; ensuring that examples
+    # can at least import, run and render something
+    if not is_lavapipe:
+        pytest.skip("screenshot comparisons are only done when using lavapipe")
+
+    # regenerate screenshot if requested
+    screenshot_path = screenshots_dir / f"{module}.png"
+    if pytestconfig.getoption("regenerate_screenshots"):
+        imageio.imwrite(screenshot_path, img)
+
+    # if a reference screenshot exists, assert it is equal
+    assert (
+        screenshot_path.exists()
+    ), "found # test_example = true but no reference screenshot available"
+    stored_img = imageio.imread(screenshot_path)
+    # assert similarity
+    is_similar = np.allclose(img, stored_img, atol=1)
+    update_diffs(module, is_similar, img, stored_img)
+    assert is_similar, (
+        f"rendered image for example {module} changed, see "
+        f"the {diffs_dir.relative_to(ROOT).as_posix()} folder"
+        " for visual diffs (you can download this folder from"
+        " CI build artifacts as well)"
+    )
+
+
+def update_diffs(module, is_similar, img, stored_img):
+    diffs_dir.mkdir(exist_ok=True)
+    diffs = {
+        diffs_dir / f"{module}-rgb.png": np.abs(stored_img[..., :3] - img[..., :3]),
```

Comment:
I suspect these should be `.astype(np.int16)`'d before subtracting them, to avoid overflow errors. Also wondering whether you want to amplify the error. If the red channel differs by 5, that'd be hard to see.

Reply:
Done!